ResourceDefaultVersion - added Primary ID TypeArn and made delete han…

ugudip opened the pull request:
1. Added Primary ID for ResourceDefaultVersion - TypeArn
2. Resolved few bugs - version conflicts and exceptions
3. Made unit test changes, jacoco coverage >80%

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

tyangjie commented on aws-cloudformation-resourcedefaultversion/aws-cloudformation-resourcedefaultversion.json:
```diff
@@ -13,6 +13,11 @@
       "pattern": "^[A-Za-z0-9]{2,64}::[A-Za-z0-9]{2,64}::[A-Za-z0-9]{2,64}$",
       "type": "string"
     },
+    "TypeArn": {
```
rename "Arn" to "TypeVersionArn"?

ugudip replied:
Yup, The Arn in line 6 actually represents the TypeversionArn, I can change that.

tyangjie commented on aws-cloudformation-resourcedefaultversion/aws-cloudformation-resourcedefaultversion.json:
```diff
@@ -13,6 +13,11 @@
       "pattern": "^[A-Za-z0-9]{2,64}::[A-Za-z0-9]{2,64}::[A-Za-z0-9]{2,64}$",
       "type": "string"
     },
+    "TypeArn": {
```
Rename "Arn" to "TypeVersionArn"?

tyangjie commented on aws-cloudformation-resourcedefaultversion/src/main/java/software/amazon/cloudformation/resourcedefaultversion/CreateHandler.java:
```diff
@@ -23,37 +17,40 @@
+        {
+          String generatedTypeArn = createTypeArn(request);
+          resourceModel.setTypeArn(generatedTypeArn);
+          return ProgressEvent.progress(resourceModel, callbackContext);
```
if type arn is null from the template input, then create handler does nothing?

ugudip replied:
The reason for writing it this way is because, we need to first generate the TypeArn and set that PrimaryId generated to the resource model back in the progress call chain. This makes cfn decide if it's an already existing resource or a new one. If TypeArn is null, we generate one and return an IN_PROGRESS event. This then calls the remaining part of the logic(setTypeDefaultVersion) and completes after a read handler is called.

tyangjie commented on aws-cloudformation-resourcedefaultversion/src/main/java/software/amazon/cloudformation/resourcedefaultversion/CreateHandler.java:
```diff
@@ -23,37 +17,40 @@
-            .translateToServiceRequest(Translator::translateToUpdateRequest) // same API call for Create + Update
-            .makeServiceCall((r, c) -> createResource(resourceModel, r, c))
-            .done(setTypeDefaultVersionResponse -> new ReadHandler().handleRequest(proxy, request, callbackContext, proxyClient, logger));
+        if(resourceModel.getTypeArn() == null)
+        {
```
Use ```StringUtils.isNotEmpty```,  and please take care of spacing/formatting

ugudip replied:
Noted.

tyangjie commented on aws-cloudformation-resourcedefaultversion/src/main/java/software/amazon/cloudformation/resourcedefaultversion/CreateHandler.java:
```diff
@@ -23,37 +17,40 @@
-            throw new CfnAlreadyExistsException(ResourceModel.TYPE_NAME, model.getPrimaryIdentifier().toString());
+    private String createTypeArn(ResourceHandlerRequest<ResourceModel> request) {
+        final ResourceModel resourceModel = request.getDesiredResourceState();
+        if(!StringUtils.isNullOrEmpty(resourceModel.getArn())){
```
Use ```StringUtils.isNotEmpty```, and spacing

benkehoe commented:
Isn't it useful for resources to have a standardized "Arn" property, so as a user, I don't have to figure out what inevitably-inconsistent naming convention every particular resource provider decided to adopt for their Arn property? (I feel the name about "Name" as a property).

ugudip commented:
I made the following changes:
1. Made "Arn" as the Physical Id for ResourceDefaultVersion which represents the overall type(ex: arn:aws:cloudformation:us-west-2:336098619110:type/resource/Organization-Service-Resource11) and added it as a readOnly property
2. Changed the current "Arn" to "TypeVersionArn" which represents the Arn of the ResourceVersion. The user should now give "TypeVersionArn" in the template to refer ResourceVersion's Arn.
3. Kept the ResourceVersion's "Arn" as is because it represents the physical Id of the ResourceVersion and makes sense to keep it Arn.
3. Changed the Unit tests and code formatting.
4. Added a check to get the latest versionID in perdictArn() - bug fix

jotompki commented on aws-cloudformation-resourcedefaultversion/src/main/java/software/amazon/cloudformation/resourcedefaultversion/CreateHandler.java:
```diff
@@ -18,42 +12,44 @@
-            .translateToServiceRequest(Translator::translateToUpdateRequest) // same API call for Create + Update
-            .makeServiceCall((r, c) -> createResource(resourceModel, r, c))
-            .done(setTypeDefaultVersionResponse -> new ReadHandler().handleRequest(proxy, request, callbackContext, proxyClient, logger));
+        if (StringUtils.isNullOrEmpty(resourceModel.getArn())) {
```
i feel like we should be using the callbackContext to set any sort of stateful things like the arn to indicate a transition. is the arn not being included in the initial input expected? what if someone includes this in their template? using the callbackContext to check if the arn has been set might avoid abusing the arn being set as some sort of state.

ugudip replied:
Made changes.

jotompki commented on aws-cloudformation-resourcedefaultversion/src/main/java/software/amazon/cloudformation/resourcedefaultversion/CreateHandler.java:
```diff
@@ -18,42 +12,44 @@
+        if (!StringUtils.isNullOrEmpty(resourceModel.getTypeVersionArn())) {
+            return resourceModel.getTypeVersionArn().substring(0, resourceModel.getTypeVersionArn().lastIndexOf("/"));
+        } else {
+            // generating Arn from the TypeName and versionID
```
versionId isn't used to generate arn.

ugudip replied:
Sure, I will change the comment here, I actually meant when TypeName and VersionId are given as an input..

jotompki commented on aws-cloudformation-resourcedefaultversion/src/main/java/software/amazon/cloudformation/resourcedefaultversion/CreateHandler.java:
```diff
@@ -18,42 +12,44 @@
+                            model.getTypeVersionArn(), model.getTypeName(), model.getVersionId()));
+                    return proxy.initiate("resourceDefaultVersion::Create", proxyClient, model, progress.getCallbackContext())
+                            .translateToServiceRequest(Translator::translateToUpdateRequest)
+                            .makeServiceCall((setTypeDefaultVersionRequest, client) -> proxyClient.injectCredentialsAndInvokeV2(setTypeDefaultVersionRequest, proxyClient.client()::setTypeDefaultVersion))
```
do we map the TypeNotFoundException that's thrown here to the notfound exception in the handler? if  a resource type doesnt exist and we try to set a default version, what happens?

ugudip replied:
In that case, the resource creation fails anyway. But I added the check in the handle error clause to let Cfn know it's a NotFoundException.
